src/messages/MOSDPGRecoveryDelete.h

```diff
@@ -5,23 +5,27 @@
 #define CEPH_MOSDPGRECOVERYDELETE_H
 
 #include "MOSDFastDispatchOp.h"
+#include "include/ceph_features.h"
 
 /*
  * instruct non-primary to remove some objects during recovery
  */
 
 struct MOSDPGRecoveryDelete : public MOSDFastDispatchOp {
 
-  static const int HEAD_VERSION = 1;
+  static const int HEAD_VERSION = 2;
   static const int COMPAT_VERSION = 1;
 
   pg_shard_t from;
   spg_t pgid;            ///< target spg_t
   epoch_t map_epoch, min_epoch;
   list<pair<hobject_t, eversion_t> > objects;    ///< objects to remove
+  uint64_t osd_features;
 
 private:
   uint64_t cost;
+  bool partial_decode_needed, final_decode_needed;
```

Comment:
I don't understand why this depends on the OSDMap.  Shouldn't it be conditional on the connection feature bits for the sending OSD?

It looks like decode_message() in Message.cc does set_connection() before calling decode_message, which means that we should be able to get the sender's feature bits from get_connection()?

Reply:
yeah, I think we can - running through upgrade suites again, though it passes my local bad-encoding=>this branch upgrade test